projects/packages/publicize/src/class-publicize-base.php

```diff
@@ -1615,6 +1708,20 @@ public function add_jetpack_social_og_images( $tags ) {
 		$opengraph_image = $this->get_social_opengraph_image( get_the_ID() );
 
 		if ( empty( $opengraph_image ) ) {
+			// If we do not have a SIG or attached image, but we have an image in post body
+			// we need to check that the image is not too large for the social sites.
+			if ( ! empty( $tags['og:image'] ) &&
+				! empty( $tags['og:image:width'] ) &&
+				! empty( $tags['og:image:height'] ) ) {
+				$reduced_image = $this->reduce_file_size( $tags['og:image'], $tags['og:image:width'], $tags['og:image:height'] );
```

Comment:
I feel a lot of the things are happening in the `reduce_file_size` function. Reduce file size should just iteratively reduce file size. Scaling down should be taken out of it.

The flow at line 1716 should perhaps be more like this:

- Get the file size and dimensions 
- Scale only if necessary. If the image fits the dimensions of a link preview, why should we scale it?
- Get file size and reduce file size, if necessary.